core/trino-spi/src/main/java/io/trino/spi/cache/PlanSignature.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi.cache;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.trino.spi.type.Type;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.StringJoiner;
+
+import static io.airlift.slice.SizeOf.estimatedSizeOf;
+import static io.airlift.slice.SizeOf.instanceSize;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Plan signature is a normalized and canonicalized representation of subplan.
+ * Plan signatures allow to identify, match and adapt similar subqueries.
+ * Concept of plan signatures is described in http://www.cs.columbia.edu/~jrzhou/pub/cse.pdf
+ */
+public class PlanSignature
+{
+    private static final int INSTANCE_SIZE = instanceSize(PlanSignature.class);
+
+    /**
+     * Key of a plan signature. Plans that can be potentially adapted
+     * to produce the same results (e.g. using column pruning, filtering or aggregation)
+     * will share the same key.
+     */
+    private final SignatureKey key;
+    /**
+     * List of group by columns if plan signature represents aggregation.
+     */
+    private final Optional<List<CacheColumnId>> groupByColumns;
+    /**
+     * List of output columns.
+     */
+    private final List<CacheColumnId> columns;
+    /**
+     * List of output columns types parallel to {@link PlanSignature#columns}.
+     */
+    private final List<Type> columnsTypes;
+
+    private volatile int hashCode;
```

Comment:
It's neither good or bad. In this case I prefer to keep it since computation of hashCode might be expensive for signature, hence I would rather do really once.

Reply:
It's unnecessary. You can get approximately the same semantics (minus the impact of a volatile read) by writing to a non-volatile int field, since there's no tearing for that type and the hashcode is deterministic (e.g., see how Java's String class does it).